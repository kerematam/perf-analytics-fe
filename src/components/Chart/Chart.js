import React, { useMemo } from 'react'
import Highcharts from 'highcharts'
import HighchartsReact from 'highcharts-react-official'

import styles from './Chart.module.scss'

const Chart = ({ title, time = [], data = [] }) => {
  const options = useMemo(
    () => ({
      title: {
        text: title,
      },
      xAxis: {
        categories: time.map(date => new Date(date).toLocaleString()),
      },
      series: [
        {
          data,
        },
      ],
    }),
    [title, data, time]
  )

  return (
    <div className={styles.chart}>
      <HighchartsReact highcharts={Highcharts} options={options} />
    </div>
  )
}

export default Chart
