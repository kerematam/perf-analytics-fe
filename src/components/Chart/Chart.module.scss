.chart {
  padding-top: 40px;
}
