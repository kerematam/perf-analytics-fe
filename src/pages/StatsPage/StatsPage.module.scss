.page_view_container {
  width: 100%;
  height: 100%;
  .top_menu {
    padding-bottom: 5px;
    display: flex;
    justify-content: flex-end;
    .button {
      margin-left: 5px;
    }
  }
  .iframe {
    width: 100%;
    height: 100%;
  }
}
