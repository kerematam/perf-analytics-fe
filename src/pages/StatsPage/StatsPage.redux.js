import { useInjectReducer, useInjectSaga } from 'redux-injectors'
import saga from './StatsPage.saga'
import { createSlice } from '@reduxjs/toolkit'

export const NAMESPACE = 'StatsPage'

export const initialState = {
  loading: {
    loadStats: false,
    getPage: false,
  },
  page: {},
  stats: [],
}

const slice = createSlice({
  name: NAMESPACE,
  initialState,
  reducers: {
    loadStats: state => {
      state.loading.loadStats = true
    },
    loadStatsSucceed: (state, action) => {
      state.stats = action.payload.stats
      state.loading.loadStats = false
    },
    loadStatsFailed: (state, action) => {
      state.stats = []
      state.loading.loadStats = false
    },

    getPage: state => {
      state.loading.getPage = true
    },
    getPageSucceed: (state, action) => {
      state.page = action.payload.page
      state.loading.getPage = false
    },
    getPageFailed: (state, action) => {
      state.state.loading.getPage = false
    },
  },
})

export const { reducer, actions } = slice

// Injectors
export const useStatsReducerInjector = () =>
  useInjectReducer({ key: NAMESPACE, reducer })
export const useStatsSagaInjector = () =>
  useInjectSaga({ key: NAMESPACE, saga })

export const useStatsReduxInjector = () => {
  useStatsReducerInjector()
  useStatsSagaInjector()
}

// Selectors
export const pageSelector = state => state[NAMESPACE]
