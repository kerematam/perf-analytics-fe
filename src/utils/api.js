import request from 'utils/request'

export const getQueryString = params =>
  Object.keys(params)
    .map(key => key + '=' + params[key])
    .join('&')

const baseURL = 'http://localhost:4040'
const baseRoute = '/api'
const commonRequest = path => `${baseURL}${baseRoute}${path}`

export const api = {
  postPage: body => {
    const requestURL = commonRequest('/sites')
    return request(requestURL, {
      body: JSON.stringify(body),
      method: 'POST',
    })
  },
  getPages: () => {
    const requestURL = commonRequest('/sites')
    return request(requestURL)
  },
  getPage: ({ pathParams }) => {
    const { pageId } = pathParams
    const requestURL = commonRequest(`/sites/${pageId}`)
    return request(requestURL)
  },
  getStats: ({ pathParams, queryParams }) => {
    const { pageId } = pathParams
    const qs = getQueryString(queryParams)
    const requestURL = commonRequest(`/sites/${pageId}/metrics/stats?${qs}`)
    return request(requestURL)
  },
}
