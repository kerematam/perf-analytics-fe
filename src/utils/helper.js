import sinon from 'sinon'

export const msg = {
  error: (err = 'Error') => {
    setTimeout(() => {
      // TODO: code proper alert component
      alert(err)
    }, 0)
  },
}

export const isTestEnv = () => process.env.NODE_ENV === 'test'

export const spy = {
  cleanerArr: [],
  resetHistory: function() {
    this.cleanerArr.forEach(fn => fn())
  },
  watch: function(fn) {
    const sandbox = sinon.createSandbox()
    sandbox.spy(fn)
    this.cleanerArr.push(sandbox.resetHistory)

    return fn
  },
}
