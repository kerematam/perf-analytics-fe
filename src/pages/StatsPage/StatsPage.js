import React, { useRef, useEffect, useMemo } from 'react'
import { Helmet } from 'react-helmet-async'
import Highcharts from 'highcharts'
import HighchartsReact from 'highcharts-react-official'
import { Button, Grid } from '@material-ui/core'
import {
  pageSelector,
  useStatsReduxInjector,
} from 'pages/StatsPage/StatsPage.redux.js'
import { useRouteMatch } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import { msg } from 'utils/helper'
import { actions } from './StatsPage.redux'

import styles from './StatsPage.module.scss'

const StatsPage = () => {
  useStatsReduxInjector()
  const dispatch = useDispatch()
  const iframeRef = useRef()
  const { page, stats } = useSelector(pageSelector)

  const {
    params: { id: pageId },
  } = useRouteMatch()

  useEffect(() => {
    dispatch(actions.getPage({ pageId }))
    dispatch(actions.loadStats({ pageId }))

    // eslint-disable-next-line react-hooks/exhaustive-deps,
  }, [])

  const handleTriggerPageLoad = () => {
    try {
      iframeRef.current.contentWindow.location.reload()
    } catch (err) {
      msg.error(err)
    }
  }

  const handleStatsLoad = () => {
    dispatch(actions.loadStats({ pageId }))
  }

  return (
    <div>
      <Helmet>
        <title>{`Stats: ${page?.url}`}</title>
      </Helmet>
      <Grid container spacing={3}>
        <Grid
          className={styles.page_view_container}
          item
          xs={12}
          sm={12}
          md={6}
        >
          <div className={styles.top_menu}>
            <Button
              onClick={handleTriggerPageLoad}
              className={styles.button}
              variant="contained"
              size="small"
            >
              Reload Page
            </Button>
            <Button
              onClick={handleStatsLoad}
              className={styles.button}
              variant="contained"
              size="small"
            >
              Reload Stats
            </Button>
          </div>
          <iframe
            src={page.url}
            title={page.url}
            className={styles.iframe}
            ref={iframeRef}
          ></iframe>
        </Grid>
        <Grid item xs={12} sm={12} md={6}>
          <Chart title="Window Load" data={stats.wl} time={stats.createdAt} />
        </Grid>
        <Grid item xs={12} sm={12} md={6}>
          <Chart
            title="Time to Fist Btye"
            data={stats.ttfb}
            time={stats.createdAt}
          />
        </Grid>
        <Grid item xs={12} sm={12} md={6}>
          <Chart
            title="First ContentFul Paint"
            data={stats.fcp}
            time={stats.createdAt}
          />
        </Grid>
        <Grid item xs={12} sm={12} md={6}>
          <Chart title="Dom Load" data={stats.doml} time={stats.createdAt} />
        </Grid>
      </Grid>
    </div>
  )
}

export const Chart = ({ title, time = [], data = [] }) => {
  const options = useMemo(
    () => ({
      title: {
        text: title,
      },
      xAxis: {
        categories: time.map(date => new Date(date).toLocaleString()),
      },
      series: [
        {
          data,
        },
      ],
    }),
    [title, data, time]
  )

  return (
    <div className={styles.chart}>
      <HighchartsReact highcharts={Highcharts} options={options} />
    </div>
  )
}

export default StatsPage
