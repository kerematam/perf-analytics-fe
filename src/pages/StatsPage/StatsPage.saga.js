import { call, put, takeLatest } from 'redux-saga/effects'
import { push } from 'connected-react-router'
import { actions } from './StatsPage.redux'
import { api } from 'utils/api'

export function* loadStats({ payload }) {
  const { pageId } = payload
  const last30Min = new Date(
    new Date().getTime() - 1000 * 60 * 30
  ).toISOString()
  try {
    const stats = yield call(api.getStats, {
      pathParams: { pageId },
      queryParams: { fromDate: last30Min },
    })
    yield put(actions.loadStatsSucceed({ stats }))
  } catch (err) {
    yield put(actions.loadStatsFailed())
  }
}

export function* getPage({ payload }) {
  const { pageId } = payload
  try {
    const page = yield call(api.getPage, { pathParams: { pageId } })
    yield put(actions.getPageSucceed({ page }))
  } catch (err) {
    yield put(push('/404'))
    yield put(actions.getPageFailed())
  }
}

export default function* charactersData() {
  yield takeLatest(actions.loadStats.type, loadStats)
  yield takeLatest(actions.getPage.type, getPage)
}
