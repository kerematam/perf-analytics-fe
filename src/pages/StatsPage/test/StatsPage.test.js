import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import configureStore from 'redux/configureStore'
import { HelmetProvider } from 'react-helmet-async'
import Adapter from 'enzyme-adapter-react-16'
import { mount, shallow, configure } from 'enzyme'
import { api } from 'utils/api'
import { spy } from 'utils/helper'
import StatsPage from 'pages/StatsPage'
import Chart from 'components/Chart'
import { sagas } from '../StatsPage.saga'
import { actions } from '../StatsPage.redux'

spy.watch(api)
spy.watch(sagas)
spy.watch(actions)

configure({ adapter: new Adapter() })

const setup = ({ renderFn, path = '/stats/5f9f3fe7dcb96300041b7ae9' }) => {
  spy.resetHistory()
  const store = configureStore()
  const Component = (
    <Provider store={store}>
      <MemoryRouter initialEntries={[path]}>
        <HelmetProvider>
          <StatsPage />
        </HelmetProvider>
      </MemoryRouter>
    </Provider>
  )

  const wrapper = renderFn(Component)

  return { wrapper, store }
}

describe('StatsPage ', () => {
  test('should match with snapshot', () => {
    const renderFn = Component => mount(Component).debug()
    const { wrapper: snapshot } = setup({ renderFn })

    expect(snapshot).toMatchSnapshot()
  })

  test('should handle redux action/state correctly', () => {
    const { store, wrapper } = setup({ renderFn: mount })
    expect(wrapper.find(Chart).length).toBe(4)

    const injectedReducers = store.injectedReducers
    const initialState = store.getState()

    expect(injectedReducers.StatsPage).toBeTruthy()
    expect(initialState.StatsPage.page).toBeTruthy()
    expect(initialState.StatsPage.stats.length).toBe(0)

    expect(actions.getPage.callCount).toBe(1)
    expect(sagas.loadStats.callCount).toBe(1)
    expect(api.getStats.callCount).toBe(1)

    const stats = {
      _id: [
        '5fa382ac6f79d10004b867dd',
        '5fa382c66f79d10004b867de',
        '5fa385a76f79d10004b867df',
      ],
      doml: [730, 518, 1015],
      ttfb: [1, 1, 1],
      fcp: [113.35499992128462, 95.41499998886138, 95.6349999178201],
      wl: [759, 555, 1052],
      createdAt: [
        '2020-11-05T04:42:20.574Z',
        '2020-11-05T04:42:46.349Z',
        '2020-11-05T04:55:03.640Z',
      ],
    }

    store.dispatch(
      actions.loadStatsSucceed({
        stats,
      })
    )

    const newState = store.getState()
    expect(newState.StatsPage.stats._id.length).toBe(3)

    const expectedDateStr = new Date(stats.createdAt[0]).toLocaleString()
    expect(wrapper.html()).toContain(expectedDateStr)
  })
})
