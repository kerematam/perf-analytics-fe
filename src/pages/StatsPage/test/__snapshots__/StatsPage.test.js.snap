// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatsPage  should match with snapshot 1`] = `
"<Provider store={{...}}>
  <MemoryRouter initialEntries={{...}}>
    <Router history={{...}}>
      <HelmetProvider context={{...}}>
        <StatsPage>
          <div>
            <Helmet defer={true} encodeSpecialCharacters={true}>
              <HelmetDispatcher defer={true} encodeSpecialCharacters={true} title=\\"Stats: undefined\\" titleAttributes={{...}} context={{...}} />
            </Helmet>
            <WithStyles(ForwardRef(Grid)) container={true} spacing={3}>
              <ForwardRef(Grid) classes={{...}} container={true} spacing={3}>
                <div className=\\"MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3\\">
                  <WithStyles(ForwardRef(Grid)) item={true} xs={12} sm={12} md={6}>
                    <ForwardRef(Grid) classes={{...}} item={true} xs={12} sm={12} md={6}>
                      <div className=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-6\\">
                        <div className=\\"page_view_container\\">
                          <div className=\\"top_menu\\">
                            <WithStyles(ForwardRef(Button)) onClick={[Function: handleTriggerPageLoad]} className=\\"button\\" variant=\\"contained\\" size=\\"small\\">
                              <ForwardRef(Button) classes={{...}} onClick={[Function: handleTriggerPageLoad]} className=\\"button\\" variant=\\"contained\\" size=\\"small\\">
                                <WithStyles(ForwardRef(ButtonBase)) className=\\"MuiButton-root MuiButton-contained button MuiButton-containedSizeSmall MuiButton-sizeSmall\\" component=\\"button\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"Mui-focusVisible\\" type=\\"button\\" onClick={[Function: handleTriggerPageLoad]}>
                                  <ForwardRef(ButtonBase) classes={{...}} className=\\"MuiButton-root MuiButton-contained button MuiButton-containedSizeSmall MuiButton-sizeSmall\\" component=\\"button\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"Mui-focusVisible\\" type=\\"button\\" onClick={[Function: handleTriggerPageLoad]}>
                                    <button className=\\"MuiButtonBase-root MuiButton-root MuiButton-contained button MuiButton-containedSizeSmall MuiButton-sizeSmall\\" onBlur={[Function]} onClick={[Function: handleTriggerPageLoad]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onDragLeave={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} tabIndex={0} type=\\"button\\" disabled={false}>
                                      <span className=\\"MuiButton-label\\">
                                        Reload Page
                                      </span>
                                      <NoSsr>
                                        <WithStyles(undefined) center={false}>
                                          <ForwardRef(TouchRipple) classes={{...}} center={false}>
                                            <span className=\\"MuiTouchRipple-root\\">
                                              <TransitionGroup component={{...}} exit={true} childFactory={[Function: childFactory]} />
                                            </span>
                                          </ForwardRef(TouchRipple)>
                                        </WithStyles(undefined)>
                                      </NoSsr>
                                    </button>
                                  </ForwardRef(ButtonBase)>
                                </WithStyles(ForwardRef(ButtonBase))>
                              </ForwardRef(Button)>
                            </WithStyles(ForwardRef(Button))>
                            <WithStyles(ForwardRef(Button)) onClick={[Function: handleStatsLoad]} className=\\"button\\" variant=\\"contained\\" size=\\"small\\">
                              <ForwardRef(Button) classes={{...}} onClick={[Function: handleStatsLoad]} className=\\"button\\" variant=\\"contained\\" size=\\"small\\">
                                <WithStyles(ForwardRef(ButtonBase)) className=\\"MuiButton-root MuiButton-contained button MuiButton-containedSizeSmall MuiButton-sizeSmall\\" component=\\"button\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"Mui-focusVisible\\" type=\\"button\\" onClick={[Function: handleStatsLoad]}>
                                  <ForwardRef(ButtonBase) classes={{...}} className=\\"MuiButton-root MuiButton-contained button MuiButton-containedSizeSmall MuiButton-sizeSmall\\" component=\\"button\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"Mui-focusVisible\\" type=\\"button\\" onClick={[Function: handleStatsLoad]}>
                                    <button className=\\"MuiButtonBase-root MuiButton-root MuiButton-contained button MuiButton-containedSizeSmall MuiButton-sizeSmall\\" onBlur={[Function]} onClick={[Function: handleStatsLoad]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onDragLeave={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} tabIndex={0} type=\\"button\\" disabled={false}>
                                      <span className=\\"MuiButton-label\\">
                                        Reload Stats
                                      </span>
                                      <NoSsr>
                                        <WithStyles(undefined) center={false}>
                                          <ForwardRef(TouchRipple) classes={{...}} center={false}>
                                            <span className=\\"MuiTouchRipple-root\\">
                                              <TransitionGroup component={{...}} exit={true} childFactory={[Function: childFactory]} />
                                            </span>
                                          </ForwardRef(TouchRipple)>
                                        </WithStyles(undefined)>
                                      </NoSsr>
                                    </button>
                                  </ForwardRef(ButtonBase)>
                                </WithStyles(ForwardRef(ButtonBase))>
                              </ForwardRef(Button)>
                            </WithStyles(ForwardRef(Button))>
                          </div>
                          <iframe src={[undefined]} title={[undefined]} className=\\"iframe\\" />
                        </div>
                      </div>
                    </ForwardRef(Grid)>
                  </WithStyles(ForwardRef(Grid))>
                  <WithStyles(ForwardRef(Grid)) item={true} xs={12} sm={12} md={6}>
                    <ForwardRef(Grid) classes={{...}} item={true} xs={12} sm={12} md={6}>
                      <div className=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-6\\">
                        <Chart title=\\"Window Load\\" data={[undefined]} time={[undefined]}>
                          <div className=\\"chart\\">
                            <ForwardRef highcharts={{...}} options={{...}}>
                              <div />
                            </ForwardRef>
                          </div>
                        </Chart>
                      </div>
                    </ForwardRef(Grid)>
                  </WithStyles(ForwardRef(Grid))>
                  <WithStyles(ForwardRef(Grid)) item={true} xs={12} sm={12} md={6}>
                    <ForwardRef(Grid) classes={{...}} item={true} xs={12} sm={12} md={6}>
                      <div className=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-6\\">
                        <Chart title=\\"Time to First Btye\\" data={[undefined]} time={[undefined]}>
                          <div className=\\"chart\\">
                            <ForwardRef highcharts={{...}} options={{...}}>
                              <div />
                            </ForwardRef>
                          </div>
                        </Chart>
                      </div>
                    </ForwardRef(Grid)>
                  </WithStyles(ForwardRef(Grid))>
                  <WithStyles(ForwardRef(Grid)) item={true} xs={12} sm={12} md={6}>
                    <ForwardRef(Grid) classes={{...}} item={true} xs={12} sm={12} md={6}>
                      <div className=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-6\\">
                        <Chart title=\\"First ContentFul Paint\\" data={[undefined]} time={[undefined]}>
                          <div className=\\"chart\\">
                            <ForwardRef highcharts={{...}} options={{...}}>
                              <div />
                            </ForwardRef>
                          </div>
                        </Chart>
                      </div>
                    </ForwardRef(Grid)>
                  </WithStyles(ForwardRef(Grid))>
                  <WithStyles(ForwardRef(Grid)) item={true} xs={12} sm={12} md={6}>
                    <ForwardRef(Grid) classes={{...}} item={true} xs={12} sm={12} md={6}>
                      <div className=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-6\\">
                        <Chart title=\\"Dom Load\\" data={[undefined]} time={[undefined]}>
                          <div className=\\"chart\\">
                            <ForwardRef highcharts={{...}} options={{...}}>
                              <div />
                            </ForwardRef>
                          </div>
                        </Chart>
                      </div>
                    </ForwardRef(Grid)>
                  </WithStyles(ForwardRef(Grid))>
                </div>
              </ForwardRef(Grid)>
            </WithStyles(ForwardRef(Grid))>
          </div>
        </StatsPage>
      </HelmetProvider>
    </Router>
  </MemoryRouter>
</Provider>"
`;
